File: web/src/finance/budgetGate.ts
// ============================================================================
// BudgetGate —— 财务 v2 B6 预算超支拦截 UI 门控纯函数（Web）
// ============================================================================
//
// 任务: stage5-finance-v2 / Task 6（预算硬约束 + 超支拦截）Web 集成层
// 路径: web/src/finance/budgetGate.ts
// 作用: 把 budgetEnforcer 的三档判定结果（OK / WARNING / BLOCK）映射为
//       UI 门控动作与零知识文案：
//         - BLOCK   → 弹“超支确认”模态，用户显式确认后方可保存；
//         - WARNING → 保存后以 toast 提示接近预算上限；
//         - OK      → 无额外门控。
//
// 铁律口径（与 Android 镜像语义一致）:
//   1. 纯函数模块：不依赖 store / 路由 / DOM / Network，输入结果对象，
//      输出布尔判定与中文文案，便于单测双端锁定；
//   2. 零知识文案：文案中只允许出现“百分比 + 分类名”，禁止金额、日期、
//      卡号、对手方等任何敏感数值；分类为 'all' 时用“全部支出”统称；
//   3. 百分比直接取 BudgetCheckResult.usedPct（bigint 整除向下取整后的
//      整数），本模块不做任何数值运算。
//
// 关联:
//   - web/src/finance/budgetEnforcer.ts（BudgetCheckResult / BudgetLevel）
//   - web/src/views/finance/BudgetConfirmDialog.vue（BLOCK 模态）
//   - web/src/views/finance/FinanceTxEditor.vue（门控调用方）
// ============================================================================

import type { BudgetCheckResult } from './budgetEnforcer'

/**
 * 是否需要弹“超支确认”模态。
 *
 * 仅 BLOCK 档需要阻断式确认；WARNING / OK 均不打断保存流程。
 */
export function needsConfirmDialog(r: BudgetCheckResult): boolean {
  return r.level === 'BLOCK'
}

/**
 * 是否需要在保存成功后给 WARNING toast。
 *
 * 仅 WARNING 档提示；BLOCK 已由确认模态承接，OK 不提示。
 */
export function needsWarningToast(r: BudgetCheckResult): boolean {
  return r.level === 'WARNING'
}

/**
 * BLOCK 模态正文文案（零知识：仅百分比，无金额 / 日期）。
 */
export function confirmText(r: BudgetCheckResult): string {
  return `预计已用 ${r.usedPct}%，超过预算阈值，是否仍保存？`
}

/**
 * WARNING toast 文案（零知识：仅百分比 + 分类名）。
 *
 * 预算分类为特殊值 'all' 时统称“全部支出”，否则展示具体分类名。
 */
export function warningText(r: BudgetCheckResult): string {
  return r.category === 'all'
    ? `本月全部支出已用 ${r.usedPct}%，接近预算上限`
    : `本月${r.category}已用 ${r.usedPct}%，接近预算上限`
}

File: android/app/src/main/java/com/everything/eve/ui/finance/BudgetGate.kt
package com.everything.eve.ui.finance

import com.everything.eve.finance.BudgetCheckResult
import com.everything.eve.finance.BudgetLevel

/**
 * B6 预算硬约束 · 零知识 UI 闸门判定。
 *
 * 纯 JVM 纯函数 / 纯文案层：只消费第一批纯函数 [BudgetEnforcer] 的判定结果
 * [BudgetCheckResult]，决定 UI 该"放行、提示、拦截确认"中的哪一种，并产出
 * 可直接展示的中文文案。
 *
 * 零知识铁律（安全红线）：
 *  - 文案只允许出现"已用百分比 + 分类名"；
 *  - 禁止出现金额（元/分/币种数字）、日期、卡号、对手方账户等任何敏感信息；
 *  - 分类为 "all" 时降级为"全部支出"，不回显原始分类串。
 *
 * 文案刻意硬编码而非走 strings.xml：纯函数 JVM 单测需对稳定字符串断言，
 * 且字符串不含任何用户数据。
 */
object BudgetGate {

    /**
     * 是否需要弹"超支确认"对话框（拦截态）。
     *
     * 仅 [BudgetLevel.BLOCK] 为 true：此时保存动作必须中止，等用户显式选择
     * "仍保存"（携带 overspendAcknowledged=true 二次提交）或"返回修改"。
     */
    fun needsConfirmDialog(result: BudgetCheckResult): Boolean =
        result.level == BudgetLevel.BLOCK

    /**
     * 是否需要在保存成功后轻提示（预警态）。
     *
     * 仅 [BudgetLevel.WARNING] 为 true：WARNING 不拦截保存，仅 Toast 提醒；
     * OK / BLOCK 均不 Toast（BLOCK 走模态对话框而非轻提示）。
     */
    fun needsWarningToast(result: BudgetCheckResult): Boolean =
        result.level == BudgetLevel.WARNING

    /**
     * 预警 Toast 文案（零知识：只含百分比与分类名）。
     *
     * 分类为 "all"（预算覆盖全部分类）时显示"全部支出"；
     * 其他分类直接展示业务分类名（纯标签，如"餐饮"），不泄露金额与日期。
     */
    fun warningText(result: BudgetCheckResult): String {
        val label = categoryLabel(result.category)
        return "本月${label}已用 ${result.usedPct}%，接近预算上限"
    }

    /**
     * 拦截确认对话框正文（零知识：只含百分比）。
     *
     * 不给分类名也成立：用户此刻正在编辑本笔流水，上下文已足够；
     * 正文刻意不出现任何金额 / 日期信息。
     */
    fun confirmText(result: BudgetCheckResult): String =
        "预计已用 ${result.usedPct}%，超过预算阈值，是否仍保存？"

    /**
     * 零知识分类标签："all" 映射为"全部支出"，其余原样返回（分类是用户自填
     * 的短标签，不含金额 / 账号等敏感字段；FinanceRecords 已限长 1..20 字）。
     */
    private fun categoryLabel(category: String): String =
        if (category == "all") "全部支出" else category
}
